design/httpproxy-conflict.md

```diff
@@ -0,0 +1,82 @@
+# HTTPProxy Conflict Resolution Design
+
+Status: Draft
+
+## Abstract
+
+Configuration errors in HTTPProxy resources can result in objects being marked as invalid and conversely will not serve traffic.
+These errors can come from invalid route configuration, missing service references, or invalid includes to other resources. 
+
+## Background
+HTTPProxy is designed as a multi-team aware Ingress, reduce annotation sprawl by providing a sensible home for configuration items.
+Additionally, HTTPProxy offers a feedback loop to users via a `Status` field on the object.
+Contour has further extended this to add `Conditions` which allows for a set of `Errors` and `Warnings` which inform users of errors in the HTTPProxy resource.
+
+When the first error or warning is detected, these fields are updated and in the status of the HTTPProxy object, however if there are more errors, they are not all shown to the user, only the first.
+There are ongoing efforts to change this logic to inform of all the errors or warnings dealing with a specific object.
+
+Contour currently invalidates the entire resource if a configuration error is encountered which can cause downtime for specific routes. 
+
+## Goals
+- Serve valid traffic configurations if portions of the spec are invalid
+
+## Non Goals
+- Surface a complete set errors or warnings for a resource (this is a different issue)
+
+## High-Level Design
+Contour will set the error or warning condition when an error is encountered, but do its best to still serve valid configurations.
+
+- If an invalid route is encountered, Contour will return a `404` HTTP status code to the requester
+- If an invalid service is encountered,Contour will return a`503` HTTP status code to the requester
+
+It's difficult to find errors and block them from breaking valid configurations due to how Contour currently processes objects.
+Contour takes a passive look to the cluster by reacting to changes after they have changed. 
+Due to this, Contour cannot block changes to resources before they are committed like an `Admission Controller` would be able to do (See alternatives: `HTTPProxyInstance`)
+
+## Detailed Design
+
+### Invalid Include
+Contour utilizes an inclusion model to allow for proper route/header delegation across namespaces and teams within a Kubernetes cluster.
+If an include does not have a matching HTTPProxy.Name or HTTPProxy.Namespace, then Contour will set an error in the Conditions of the object.
```

Comment:
If we report an error in the conditions, then will the HTTPProxy be marked as invalid, i.e. the `Valid` condition would be set to `false`? To me, that's a little misleading, since parts of it are valid and being programmed with Envoy. It seems like we would need to move to a non-binary model of validity in the status.

Reply:
Yes, the current behavior is that there is one Condition (`Valid`), that is `status: true` as long as there are *zero* errors in the `errors` list. What this change seems to imply to me is that missing routes or missing services would become `warnings`, and serve status codes instead, *without* impacting the validity of the HTTPProxy. Note that it's not possible to mark a condition as partially true - the only valid values for `status` are `true`, `false`, and `unknown` (which is set by convention to mean "not reconciled by owning controller yet".)

We *can* extend the meaning of the top-level validity field to summarise things in a different way - it currently may be `valid`, `invalid`, or `orphaned`. We could add another summary entry to *that* field that would mean `valid, but with problems`. That does muddy up the principle that the Conditions are the place to look, but may be an acceptable compromise while the tooling for Conditions is still not as good as the tooling for showing a single field in the `kubectl get` output.

Reply:
It feels to me that either it's working or it not, meaning, there's not a case where if Contour finds an issue, you're going to ignore it and move on. 

It might be too strict to define the entire HTTPProxy object as "valid, invalid, or orphaned". But it does give signal to the user that something is wrong and it needs to be investigated.

Reply:
It'd be nice to be able to differentiate between "the proxy was not configured at all in Envoy" vs. "the proxy was partially configured in Envoy, with these exceptions" vs. "the proxy was fully configured in Envoy".

Reply:
Resolving this thread b/c we've continued discussion on this topic in a separate thread.